File: src/Database/Python/Tests/compound_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Laboratory of Physical Chemistry, Reiher Group.
See LICENSE.txt for details.
"""

import scine_utilities as utils
import scine_database as db
import unittest
import os


class CompoundTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP') or '127.0.0.1'
        self.manager.credentials.database_name = "unittest_db_CompoundTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_creation_one(self):
        coll = self.manager.get_collection("compounds")
        compound = db.Compound.make([db.ID(), db.ID()], coll)
        assert compound.has_id()

    def test_creation_two(self):
        coll = self.manager.get_collection("compounds")
        compound = db.Compound()
        compound.link(coll)
        comp = compound.create([db.ID()])
        assert compound.get_id() == comp

    def test_create_fails_collection(self):
        comp = db.Compound()
        self.assertRaises(RuntimeError, lambda: comp.create([db.ID()]))

    def test_centroid(self):
        coll = self.manager.get_collection("compounds")
        ref = db.ID()
        compound = db.Compound.make([ref, db.ID()], coll)
        assert ref == compound.get_centroid()

    def test_centroid_fails_collection(self):
        comp = db.Compound(db.ID())
        self.assertRaises(RuntimeError, lambda: comp.get_centroid())

    def test_centroid_fails_id(self):
        coll = self.manager.get_collection("compounds")
        comp = db.Compound()
        comp.link(coll)
        self.assertRaises(RuntimeError, lambda: comp.get_centroid())

    def test_centroid_fails_structures(self):
        coll = self.manager.get_collection("compounds")
        comp = db.Compound.make([], coll)
        self.assertRaises(RuntimeError, lambda: comp.get_centroid())

    def test_reactions(self):
        s1 = db.ID()
        s2 = db.ID()
        s3 = db.ID()
        coll = self.manager.get_collection("compounds")
        comp = db.Compound.make([s1, s2, s3], coll)
        assert comp.has_id()

        # Structure Functionalities
        r1 = db.ID()
        r2 = db.ID()
        r3 = db.ID()
        r4 = db.ID()
        r5 = db.ID()
        r6 = db.ID()

        reactions = comp.get_reactions()
        assert len(reactions) == 0
        comp.set_reactions([r1, r2, r3])
        reactions = comp.get_reactions()
        assert len(reactions) == 3
        assert reactions[0] == r1
        assert reactions[1] == r2
        assert reactions[2] == r3

        comp.add_reaction(r4)
        comp.remove_reaction(r1)
        reactions = comp.get_reactions()
        assert len(reactions) == 3
        assert reactions[0] == r2
        assert reactions[1] == r3
        assert reactions[2] == r4

        comp.set_reactions([r4, r5, r6])
        reactions = comp.get_reactions()
        assert len(reactions) == 3
        assert reactions[0] == r4
        assert reactions[1] == r5
        assert reactions[2] == r6
        assert comp.has_reaction(r4)

        comp.clear_reactions()
        reactions = comp.get_reactions()
        assert len(reactions) == 0
        assert not comp.has_reaction(r4)

    def test_reactions_fails_collection(self):
        comp = db.Compound()
        self.assertRaises(RuntimeError, lambda: comp.add_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.has_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.remove_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.set_reactions([]))
        self.assertRaises(RuntimeError, lambda: comp.get_reactions())
        self.assertRaises(RuntimeError, lambda: comp.clear_reactions())

    def test_reactions_fails_id(self):
        coll = self.manager.get_collection("compounds")
        comp = db.Compound()
        comp.link(coll)
        self.assertRaises(RuntimeError, lambda: comp.add_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.has_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.remove_reaction(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.set_reactions([]))
        self.assertRaises(RuntimeError, lambda: comp.get_reactions())
        self.assertRaises(RuntimeError, lambda: comp.clear_reactions())

    def test_structures(self):
        s1 = db.ID()
        s2 = db.ID()
        s3 = db.ID()
        coll = self.manager.get_collection("compounds")
        comp = db.Compound.make([s1, s2, s3], coll)
        assert comp.has_id()

        # Structure Functionalities
        s4 = db.ID()
        s5 = db.ID()
        s6 = db.ID()

        structures = comp.get_structures()
        assert len(structures) == 3
        assert structures[0] == s1
        assert structures[1] == s2
        assert structures[2] == s3

        comp.add_structure(s4)
        comp.remove_structure(s1)
        structures = comp.get_structures()
        assert len(structures) == 3
        assert structures[0] == s2
        assert structures[1] == s3
        assert structures[2] == s4

        comp.set_structures([s4, s5, s6])
        structures = comp.get_structures()
        assert len(structures) == 3
        assert structures[0] == s4
        assert structures[1] == s5
        assert structures[2] == s6
        assert comp.has_structure(s4)

        comp.clear_structures()
        structures = comp.get_structures()
        assert len(structures) == 0
        assert not comp.has_structure(s4)

    def test_structures_fails_collection(self):
        comp = db.Compound()
        self.assertRaises(RuntimeError, lambda: comp.add_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.has_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.remove_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.set_structures([]))
        self.assertRaises(RuntimeError, lambda: comp.get_structures())
        self.assertRaises(RuntimeError, lambda: comp.clear_structures())

    def test_structures_fails_id(self):
        coll = self.manager.get_collection("compounds")
        comp = db.Compound()
        comp.link(coll)
        self.assertRaises(RuntimeError, lambda: comp.add_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.has_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.remove_structure(db.ID()))
        self.assertRaises(RuntimeError, lambda: comp.set_structures([]))
        self.assertRaises(RuntimeError, lambda: comp.get_structures())
        self.assertRaises(RuntimeError, lambda: comp.clear_structures())

File: src/Database/Tests/CompoundTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Laboratory of Physical Chemistry, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Collection.h>
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/Compound.h>
#include <Database/Objects/Structure.h>
#include <Utils/Geometry/ElementTypes.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class CompoundTest CompoundTest.cpp
 * @brief Comprises tests for the class Scine::Database::Compound.
 * @test
 */
class CompoundTest : public Test {
 public:
  CompoundTest() {
  }
  static Credentials credentials;
  static Manager db;

  static void SetUpTestCase() {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  static void TearDownTestCase() {
    db.wipe();
  }
};
Credentials CompoundTest::credentials(TEST_MONGO_DB_IP, 27017, "unittest_db_CompoundTest");
Manager CompoundTest::db;

TEST_F(CompoundTest, Create) {
  auto coll = db.getCollection("compounds");
  ID id;
  Compound compound;
  compound.link(coll);
  ID comp = compound.create({id});
  ASSERT_EQ(compound.id(), comp);
}

TEST_F(CompoundTest, CreateFails) {
  ID id;
  Compound compound;
  ASSERT_THROW(compound.create({id}), Exceptions::MissingLinkedCollectionException);
}

TEST_F(CompoundTest, Centroid) {
  auto coll = db.getCollection("compounds");
  ID id;
  Compound compound = Compound::create({id}, coll);
  ASSERT_EQ(compound.getCentroid(), id);
}

TEST_F(CompoundTest, CentroidFails1) {
  ID id;
  Compound compound;
  ASSERT_THROW(compound.getCentroid(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(CompoundTest, CentroidFails2) {
  auto coll = db.getCollection("compounds");
  ID id;
  Compound compound;
  compound.link(coll);
  ASSERT_THROW(compound.getCentroid(), Exceptions::MissingIDException);
}

TEST_F(CompoundTest, CentroidFails3) {
  auto coll = db.getCollection("compounds");
  Compound compound = Compound::create({}, coll);
  ASSERT_THROW(compound.getCentroid(), Exceptions::MissingIdOrField);
}

TEST_F(CompoundTest, Reaction) {
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3, id4, id5;
  Compound compound = Compound::create({id1}, coll);
  // Checks
  ASSERT_FALSE(compound.hasReaction(id1));
  compound.addReaction(id2);
  ASSERT_TRUE(compound.hasReaction(id2));
  ASSERT_EQ(1, compound.hasReactions());
  compound.setReactions({id3, id4, id5});
  ASSERT_EQ(3, compound.hasReactions());
  auto ret = compound.getReactions();
  ASSERT_EQ(ret[0], id3);
  ASSERT_EQ(ret[1], id4);
  ASSERT_EQ(ret[2], id5);
  compound.removeReaction(id4);
  ASSERT_EQ(2, compound.hasReactions());
  ASSERT_FALSE(compound.hasReaction(id4));
  compound.clearReactions();
  ASSERT_EQ(0, compound.hasReactions());
}

TEST_F(CompoundTest, ReactionFails1) {
  Compound compound;
  ID id;
  ASSERT_THROW(compound.hasReaction(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.hasReactions(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.getReactions(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.addReaction(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.setReactions({}), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.removeReaction(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.clearReactions(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(CompoundTest, ReactionFails2) {
  auto coll = db.getCollection("compounds");
  Compound compound;
  compound.link(coll);
  ID id;
  ASSERT_THROW(compound.hasReaction(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.hasReactions(), Exceptions::MissingIDException);
  ASSERT_THROW(compound.getReactions(), Exceptions::MissingIDException);
  ASSERT_THROW(compound.addReaction(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.setReactions({}), Exceptions::MissingIDException);
  ASSERT_THROW(compound.removeReaction(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.clearReactions(), Exceptions::MissingIDException);
}

TEST_F(CompoundTest, Structure) {
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3, id4, id5;
  Compound compound = Compound::create({id1}, coll);
  // Checks
  ASSERT_TRUE(compound.hasStructure(id1));
  compound.addStructure(id2);
  ASSERT_TRUE(compound.hasStructure(id2));
  ASSERT_EQ(2, compound.hasStructures());
  compound.setStructures({id3, id4, id5});
  ASSERT_EQ(3, compound.hasStructures());
  auto ret = compound.getStructures();
  ASSERT_EQ(ret[0], id3);
  ASSERT_EQ(ret[1], id4);
  ASSERT_EQ(ret[2], id5);
  compound.removeStructure(id4);
  ASSERT_EQ(2, compound.hasStructures());
  ASSERT_FALSE(compound.hasStructure(id4));
  compound.clearStructures();
  ASSERT_EQ(0, compound.hasStructures());
}

TEST_F(CompoundTest, StructureFails1) {
  Compound compound;
  ID id;
  ASSERT_THROW(compound.hasStructure(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.hasStructures(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.getStructures(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.addStructure(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.setStructures({}), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.removeStructure(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(compound.clearStructures(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(CompoundTest, StructureFails2) {
  auto coll = db.getCollection("compounds");
  Compound compound;
  compound.link(coll);
  ID id;
  ASSERT_THROW(compound.hasStructure(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.hasStructures(), Exceptions::MissingIDException);
  ASSERT_THROW(compound.getStructures(), Exceptions::MissingIDException);
  ASSERT_THROW(compound.addStructure(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.setStructures({}), Exceptions::MissingIDException);
  ASSERT_THROW(compound.removeStructure(id), Exceptions::MissingIDException);
  ASSERT_THROW(compound.clearStructures(), Exceptions::MissingIDException);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */
